packages/bun-vscode/src/features/tests/bun-test-controller.ts

```diff
@@ -0,0 +1,1325 @@
+      case "expect(received).toBe(expected)": {
+        const regex = /^Expected:\s*([\s\S]*?)\nReceived:\s*([\s\S]*?)$/;
+        let testMessage = vscode.TestMessage.diff(
+          errorLine,
+          messageLines.match(regex)?.[1].trim() || "",
+          messageLines.match(regex)?.[2].trim() || "",
+        );
+        if (!messageLines.match(regex)) {
+          const code = messageLines
+            .replace(/(?:\r?\n)+(- Expected\s+- \d+|\+ Received\s+\+ \d+)\s*$/g, "")
+            .replace(/(?:\r?\n)+(- Expected\s+- \d+|\+ Received\s+\+ \d+)\s*$/g, "")
+            .trim();
+          testMessage = new vscode.TestMessage(
+            new vscode.MarkdownString("Values did not match:\n").appendCodeblock(code, "diff"),
+          );
+        }
+        testMessage.location = location;
+        return testMessage;
+      }
+
+      case "expect(received).toBeInstanceOf(expected)": {
+        const regex = /^Expected constructor:\s*([\s\S]*?)\nReceived value:\s*([\s\S]*?)$/;
+        let testMessage = vscode.TestMessage.diff(
+          errorLine,
+          messageLines.match(regex)?.[1].trim() || "",
+          messageLines.match(regex)?.[2].trim() || "",
+        );
+        if (!messageLines.match(regex)) {
+          testMessage = new vscode.TestMessage(messageLines);
+        }
+        testMessage.location = location;
+        return testMessage;
+      }
+
+      case "expect(received).not.toBe(expected)":
+      case "expect(received).not.toEqual(expected)": {
+        const testMessage = new vscode.TestMessage(messageLines);
+        testMessage.location = location;
+        return testMessage;
+      }
+
+      case "expect(received).toBeNull()": {
+        const actualValue = messageLines.replace("Received:", "").trim();
+        const testMessage = vscode.TestMessage.diff(errorLine, "null", actualValue);
+        testMessage.location = location;
+        return testMessage;
+      }
+
+      case "expect(received).toMatchObject(expected)": {
+        const line = messageLines
+          .replace(/(?:\r?\n)+(- Expected\s+- \d+|\+ Received\s+\+ \d+)\s*$/g, "")
+          .replace(/(?:\r?\n)+(- Expected\s+- \d+|\+ Received\s+\+ \d+)\s*$/g, "");
+
+        const formatted = new vscode.MarkdownString("Values did not match:");
+        formatted.appendCodeblock(line, "diff");
+        const testMessage = new vscode.TestMessage(formatted);
+        testMessage.location = location;
+        return testMessage;
+      }
+    }
+
+    let lastEffortMsg = messageLines.split("\n");
+    const lastLine = lastEffortMsg?.at(-1);
+    if (lastLine?.startsWith("Received ") || lastLine?.startsWith("Received: ")) {
+      lastEffortMsg = lastEffortMsg.reverse();
+    }
+
+    const msg = errorLine.startsWith("error: expect")
+      ? `${lastEffortMsg.join("\n")}\n${errorLine.trim()}`.trim()
+      : `${errorLine.trim()}\n${messageLines}`.trim();
+
+    const testMessage = new vscode.TestMessage(msg);
```

Comment:
This is going to be brittle

Reply:
Yeah, however showing the `expect()...` as first line is a little useless as it's what user sees without opening thr popup. And then some of my other cases are for using the inbuilt diff one that shows a nice diff for user.

Maybe one day we can make socket reporter give a json or more structured output to rely on instead.